tree/tree/src/InternalTreeUtils.cxx

```diff
@@ -272,10 +317,52 @@ RFriendInfo GetFriendInfo(const TTree &tree)
             throw std::runtime_error("A TTree in the list of friends is not linked to any file. "
                                      "Friends with no associated files are not supported.");
          fileNames.emplace_back(f->GetName());
+         // We already have a pointer to the file and the tree, we can get the
+         // entries without triggering a re-open
+         entries.emplace_back(frTree->GetEntries());
+      }
+   }
+
+   return RFriendInfo{std::move(friendNames), std::move(friendFileNames), std::move(friendChainSubNames),
+                      std::move(friendEntries)};
+}
+
+std::vector<std::unique_ptr<TChain>> MakeFriendsAndAddToTree(TTree &tree, const RFriendInfo &finfo)
+{
+
+   std::vector<std::unique_ptr<TChain>> friends;
+   const auto nFriends = finfo.fFriendNames.size();
+   friends.reserve(nFriends);
+
+   for (std::size_t i = 0u; i < nFriends; ++i) {
+      const auto &thisFriendNameAlias = finfo.fFriendNames[i];
+      const auto &thisFriendName = thisFriendNameAlias.first;
+      const auto &thisFriendAlias = thisFriendNameAlias.second;
+      const auto &thisFriendFileNames = finfo.fFriendFileNames[i];
+      const auto &thisFriendChainSubNames = finfo.fFriendChainSubNames[i];
+      const auto &thisFriendEntries = finfo.fFriendEntries[i].empty()
+                                         ? std::vector<Long64_t>(thisFriendFileNames.size(), TTree::kMaxEntries)
+                                         : finfo.fFriendEntries[i];
+
+      // Build a friend chain
+      auto frChain = std::make_unique<TChain>(thisFriendName.c_str(), "", TChain::kWithoutGlobalRegistration);
+      if (thisFriendChainSubNames.empty()) {
+         // The friend is a TTree. It's safe to add to the chain the filename directly.
+         frChain->Add(thisFriendFileNames[0].c_str(), thisFriendEntries[0]);
+      } else {
+         // Otherwise, the new friend chain needs to be built using the nomenclature
+         // "filename?#treename" as argument to `TChain::Add`
+         for (std::size_t j = 0u; j < thisFriendFileNames.size(); ++j) {
+            frChain->Add((thisFriendFileNames[j] + "?#" + thisFriendChainSubNames[j]).c_str(), thisFriendEntries[j]);
```

Comment:
any reason why you don't use `?#` in all cases?

Reply:
Because `thisFriendChainSubNames` might be empty, so I can't call `thisFriendChainSubNames[0]` in case the friend is a TTree. In a separate PR we could make this pattern more homogeneous by defaulting `thisFriendChainSubNames` to always have at least one empty string.